Add API design document

mhsmith opened the pull request:
@ntoll asked me to summarize the principles we follow for naming things in Toga, so I added an API design page to the documentation.

freakboy3742 commented on docs/how-to/topics/api-design.rst:
```diff
@@ -0,0 +1,50 @@
+``value``, and the corresponding event is called ``on_change``.
+
+When a widget has a non-editable caption, (e.g. a :any:`Button` or :any:`Switch`), then
+its property is called ``text``.
```
Other notable common names: 
* `min`/`max` as expressions of range, and `step` if there are intervals.

Should probably also mention that naming follows US English spelling (grumble...)

mhsmith replied:
Done, except for `step`, which is only used in one widget (`NumberInput`), so isn't worth stating as a general rule.

freakboy3742 commented on docs/how-to/topics/api-design.rst:
```diff
@@ -0,0 +1,50 @@
+==========
+
+Toga prefers to expose an object's state using property notation (e.g.
+``widget.property``) rather than getter or setter methods.
```
Should probably mention that properties should follow [Postel's Law](https://en.wikipedia.org/wiki/Robustness_principle) - if a widget stores text, any input should be coerced into a string when set, and be guaranteed to return text when retrieved.

freakboy3742 replied:
Also -uses as setter *wherever possible*. `WebView.set_content()` is a notable example, because it requires a URL as well as the HTML content.

mhsmith replied:
Done.

freakboy3742 commented on docs/how-to/topics/api-design.rst:
```diff
@@ -0,0 +1,50 @@
+
+Events are used to notify your app of user actions. To make your app handle an event,
+you can assign either a regular or async callable to an event handler property. These
+can be identified by their names, which always begin with ``on_``.
```
Also - event names (when they're not a common name) attempt to capture the "high level purpose", not the specific mode of interaction. For example:

* `on_press`, not `on_click` on buttons - "click" implies a mouse is used.
* `on_primary_action` and `on_secondary_action`, not "left click" and "right click"; again, the action might be a swipe, or a click, or a double click.

freakboy3742 reviewed (changes requested):
Looks good - a couple of suggestions for areas that could be added/clarified.

A couple of other high-level topics worth mentioning as an API design guide:

* We write "Pythonic" APIs. We're wrapping native APIs, which means non-Python APIs. We try to expose a Pythonic API wherever possible, including using Python language idiom (e.g., using context mangers and iterators) and naming conventions (e.g., snake_case, not CamelCase), even if the underlying platforms expose APIs that don't lean that way.
* We follow the 80% rule for abstractions - the goal is to provide an API that makes 80% of use cases possible, with enough internals exposed so that advanced users can get to the other 20% if they need to. We don't consider 100% coverage of all possible features to be desirable (or even achievable in many cases). 
* We aim for adherence to native look-and-feel first. If a native platform handles a feature in a particular way, that's how we surface the feature to end users. macOS menus are app-wide at the top of the screen; Windows menus are window-specific, and in-window. This often means abstracting the "bigger picture", rather than the specific manifestation of a feature.

mhsmith commented:
> * We write "Pythonic" APIs. [...]

Added.

> * We aim for adherence to native look-and-feel first. If a native platform handles a feature in a particular way, that's how we surface the feature to end users. macOS menus are app-wide at the top of the screen; Windows menus are window-specific, and in-window. This often means abstracting the "bigger picture", rather than the specific manifestation of a feature.

I think this is already adequately covered by the "philosophy" page.

> * We follow the 80% rule for abstractions - the goal is to provide an API that makes 80% of use cases possible, with enough internals exposed so that advanced users can get to the other 20% if they need to. We don't consider 100% coverage of all possible features to be desirable (or even achievable in many cases).

I think this also belongs on the "philosophy" page, but it wasn't previously there, so I've added it.

freakboy3742 reviewed (approved):
Looks good. I'm sure we'll make plenty of updates and clarifications over time, but this is a good v1.
